src/abstractdataframe/reshape.jl

```diff
@@ -121,169 +198,212 @@ function _stackview(df::AbstractDataFrame, measure_vars::AbstractVector{Int},
-    # we have to handle a case with missings in refkeycol as levs will skip missing
-    col = similar(df[!, rowkey], length(levs) + hadmissing)
-    copyto!(col, levs)
-    hadmissing && (col[end] = missing)
-    df2 = DataFrame(unstacked_val, Symbol.(renamecols.(levels(keycol))), copycols=false)
-    return insertcols!(df2, 1, _names(df)[rowkey] => col)
-end
-
 function unstack(df::AbstractDataFrame, rowkeys, colkey::ColumnIndex,
-                 value::ColumnIndex; renamecols::Function=identity)
-    rowkey_ints = index(df)[rowkeys]
+                 value::ColumnIndex; renamecols::Function=identity,
+                 allowmissing::Bool=false, allowduplicates::Bool=false)
+    rowkey_ints = vcat(index(df)[rowkeys])
     @assert rowkey_ints isa AbstractVector{Int}
     length(rowkey_ints) == 0 && throw(ArgumentError("No key column found"))
-    length(rowkey_ints) == 1 && return unstack(df, rowkey_ints[1], colkey, value,
-                                               renamecols=renamecols)
-    g = groupby(df, rowkey_ints, sort=true)
-    keycol = categorical(df[!, colkey])
-    droplevels!(keycol)
+    g_rowkey = groupby(df, rowkey_ints)
+    g_colkey = groupby(df, colkey)
     valuecol = df[!, value]
-    return _unstack(df, rowkey_ints, index(df)[colkey], keycol, valuecol, g, renamecols)
+    return _unstack(df, rowkey_ints, index(df)[colkey], g_colkey,
+                    valuecol, g_rowkey, renamecols, allowmissing, allowduplicates)
 end
 
 function unstack(df::AbstractDataFrame, colkey::ColumnIndex, value::ColumnIndex;
-                 renamecols::Function=identity)
+                 renamecols::Function=identity,
+                 allowmissing::Bool=false, allowduplicates::Bool=false)
     colkey_int = index(df)[colkey]
     value_int = index(df)[value]
     return unstack(df, Not(colkey_int, value_int), colkey_int, value_int,
-            renamecols=renamecols)
+            renamecols=renamecols, allowmissing=allowmissing,
+            allowduplicates=allowduplicates)
 end
 
-unstack(df::AbstractDataFrame; renamecols::Function=identity) =
-    unstack(df, :variable, :value, renamecols=renamecols)
+unstack(df::AbstractDataFrame; renamecols::Function=identity,
+        allowmissing::Bool=false, allowduplicates::Bool=false) =
+    unstack(df, :variable, :value, renamecols=renamecols, allowmissing=allowmissing,
+            allowduplicates=allowduplicates)
 
-function _unstack(df::AbstractDataFrame, rowkeys::AbstractVector{Int},
-                  colkey::Int, keycol::CategoricalVector,
-                  valuecol::AbstractVector, g::GroupedDataFrame,
-                  renamecols::Function)
-    idx, starts, ends = g.idx, g.starts, g.ends
+function getrefs(g::GroupedDataFrame)
+    idx::Vector{Int}, starts::Vector{Int}, ends::Vector{Int} = g.idx, g.starts, g.ends
     groupidxs = [idx[starts[i]:ends[i]] for i in 1:length(starts)]
-    rowkey = zeros(Int, size(df, 1))
+    ref = zeros(Int, size(parent(g), 1))
```

Comment:
Why not this?
```suggestion
    ref = zeros(Int, length(groupidxs))
```
Iterating over `starts` and `ends` directly to avoid storing the intermediate `groupidxs` would make sense.

But isn't this just recomputing the `gd.groups` field? What am I missing?

Reply:
I will check - this was the original code I reused